File: dsa/circular_queue.py
from typing import Any

class Queue:
    def __init__(self, maxlen:int):
        self.front = 0
        self.rear = -1
        self.counts = 0
        self.maxlen = maxlen

        self.que:list[Any] = [0 for _ in range(maxlen)]

    def push(self, el:Any):
        if self.counts == self.maxlen:
            raise RuntimeError("Queue is full")
        self.rear = (self.rear + 1) % self.maxlen
        self.counts += 1
        self.que[self.rear] = el

    def pop(self):
        if self.counts <= 0:
            raise RuntimeError("Queue is empty")
        el = self.que[self.front]
        self.front = (self.front + 1) % self.maxlen
        self.counts -= 1
        return el
    
    def __repr__(self) -> str:
        # disp_ls = []
        # i = self.front
        # c = self.counts
        # while True:
        #     i = (i+1)%self.maxlen; c-=1
        #     disp_ls.append(self.que[i])
        #     if c<=0:
        #         break
        return f"Queue({self.que}, front={self.front}, rear={self.rear})"

if __name__ == "__main__":
    que = Queue(maxlen=4)
    que.push(10); que.push(20); que.push(30); que.push(40)
    print(que)
    try: que.push(50)
    except Exception as e: print(e)
    print(que)
    que.pop()
    print(que)
    que.push(50)
    print(que)
    

File: dsa/circular_queue.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

typedef struct {
    int front;
    int rear;
    int maxlen;
    int counts;
    int *que;
} Queue;

Queue* init(int maxlen){
    Queue* que;
    que->front = 0;
    que->rear = -1;
    que->counts = 0;
    que->maxlen = maxlen;
    que->que = (int *)malloc(maxlen*sizeof(int));
    return que;
}

void del(Queue* que){
    free(que->que);
}

void push(Queue* que, int el){
    assert(que->counts != que->maxlen);
    que->rear = (que->rear + 1) % que->maxlen;
    que->counts += 1;
    que->que[que->rear] = el;
}

int pop(Queue* que){
    assert(que->counts > 0);
    int el = que->que[que->front];
    que->front = (que->front + 1) % que->maxlen;
    que->counts -= 1;
}

void display(Queue* que){
    printf("Queue([");
    for(int i=0; i<que->counts; i++){
        printf("%i, ", que->que[i]);
    }
    printf("], front=%i, rear=%i, counts=%i)\n", que->front, que->rear, que->counts);
}

int main(){
    Queue* que = init(4);
    push(que, 10); push(que, 20); push(que, 30); push(que, 40);
    display(que); // [10, 20, 30, 40]
    int el = pop(que); 
    display(que); // [10, 20, 30]
    push(que, 50);
    display(que);
    del(que);
}